exonum-java-binding/core/src/main/java/com/exonum/binding/core/transaction/Preconditions.java

```diff
@@ -0,0 +1,267 @@
+
+import static com.google.common.base.Strings.lenientFormat;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Utility methods that helps verifying conditions conducted in expression
+ * while transaction execution.
+ * If the condition is not met, the {@code Preconditions} method throws {@link ExecutionException}.
+ *
+ * <p>Consider the following example:
+ * <pre>{@code
+ *   void checkEnoughMoney(long balance, long amount) {
+ *     if(balance < amount) {
+ *       throw new ExecutionException((byte)3, "Not enough money. Operation amount is " + amount
+ *       + ", but actual balance was " + balance);
+ *     }
+ *   }
+ * }</pre>
+ *
+ * <p>which can be replaced using Preconditions:
+ * <pre>{@code
+ *   check(amount <= balance, (byte)3,
+ *       "Not enough money. Operation amount is %s, but actual balance was %s",
+ *       amount, balance);
+ * }</pre>
+ */
+public final class Preconditions {
+
+  /**
+   * Verifies the truth of the given expression.
+   *
+   * @param expression a boolean expression
+   * @param errorCode execution error code
+   * @throws ExecutionException if {@code expression} is false
+   */
+  public static void check(boolean expression, byte errorCode) {
+    if (!expression) {
+      throw new ExecutionException(errorCode);
+    }
+  }
+
+  /**
+   * Verifies the truth of the given expression.
+   *
+   * @param expression a boolean expression
+   * @param errorCode execution error code
+   * @param errorMessage execution error description to use if the check fails
+   * @throws ExecutionException if {@code expression} is false
+   */
+  public static void check(boolean expression, byte errorCode, @Nullable Object errorMessage) {
+    if (!expression) {
+      throw new ExecutionException(errorCode, String.valueOf(errorMessage));
+    }
+  }
+
+  /**
+   * Verifies the truth of the given expression.
+   *
+   * @param expression a boolean expression
+   * @param errorCode execution error code
+   * @param errorMessageTemplate execution error description template to use if the check fails.
+   *        The template could have placeholders {@code %s} which will be replaces by arguments
+   *        resolved by position
+   * @param errorMessageArgs arguments to be used in the template. Each argument will be converted
+   *        to string using {@link String#valueOf(Object)}
+   * @throws ExecutionException if {@code expression} is false
+   */
+  public static void check(boolean expression, byte errorCode,
+      @Nullable String errorMessageTemplate,
+      @Nullable Object... errorMessageArgs) {
+    if (!expression) {
+      throw new ExecutionException(errorCode,
+          lenientFormat(errorMessageTemplate, errorMessageArgs));
+    }
+  }
+
+  /**
+   * Verifies the truth of the given expression.
+   *
+   * <p>See {@link #check(boolean, byte, String, Object...)} for details.
+   */
+  public static void check(boolean expression, byte errorCode,
```

Comment:
How the set of overloads was choosen?

Reply:
A fascinating talk on how they do it at Google, helping design not only APIs but languages: https://www.youtube.com/watch?v=sPW2Pz2dI9E&list=LLXsc0bWTsranC6H8z1IOoXg&index=10&t=0s (@vitvakatu , @slowli )

That's what we often miss — data on the actual usages to make informed decisions, instead of speculations 🙃

Reply:
just tried to cover most frequently used cases + inspired by guava lib 🙃